### ISSUE TITLE
Segmentation fault for fp8 and thunder_cudnn

### ISSUE THREAD

**Issue by mpatel31415** [Lightning-AI/lightning-thunder#756]
## 🐛 Bug

For a few models ( Platypus-30B with FSDP zero3, Gemma7b with DDP and vicuna-33b-v1.3 with FSDP zero3) we get segmentation fault error when trying to use fp8 with thunder_cudnn. When using thunder_cudnn with bf16 only we get OOM error (tested on Gemma7b).

### To Reproduce

Please use:
2 nodes, each with 8 GPUs.
Image: "INTERNAL_IMAGE:pjnl-20240705"

Training script:
python /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py \
    --model_name Gemma-7b \
    --distributed_mode ddp \
    --shard_mode None \
    --compile thunder_cudnn \
    --checkpoint_activations False \
    --low_precision_mode fp8-delayed-te \
    --micro_batch_size 1"

### Expected behavior

We should not get segmentation fault error, but at least OOM error. This issue might be related to FP8 purely, but due to [this issue](https://github.com/Lightning-AI/pytorch-lightning/issues/19989) I'm not able to verify it for eager mode. 

### Environment

system.device_product_name               DGXH100
system.gpu_driver_version                    535.129.03

libraries.cuda                               12.5.1.007

libraries.pip.lightning               2.3.0.dev20240428
libraries.pip.lightning-thunder              0.2.0.dev0
libraries.pip.lightning-utilities          0.11.3.post0
libraries.pip.litgpt                              0.3.1
libraries.pip.nvfuser                  0.2.6+gitbad998a
libraries.pip.pytorch-lightning                   2.3.2
libraries.pip.torch                  2.5.0a0+git57d05f2
libraries.pip.torchmetrics                  1.4.0.post0
libraries.pip.torchvision              0.19.0a0+d23a6e1"

### Additional context

I'm attaching full output as a file. 
[output_to_report.txt](https://github.com/user-attachments/files/16175318/output_to_report.txt)

**Comment by tfogal** [comment]
@vedaanta it sounds like we don't have enough memory to run these models, so we can't expect this to *work*.

But we shouldn't be crashing, and for some reason it seems we only crash when the cuDNN executor is enabled. Could you follow up here?

**Comment by mpatel31415** [comment]
Hi! So this issue was present recently in 1 case, below are reproduction instructions:


```
Please use:
2 node(s), each with 8 GPUs.
Image ""INTERNAL_IMAGE:pjnl-20241011""
Training script:
python /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py \
    --model_name tiny-llama-1.1b \
    --distributed_mode ddp \
    --shard_mode None \
    --compile dynamo_thunder \
    --checkpoint_activations False \
    --low_precision_mode fp8-delayed-te  \
    --micro_batch_size 18
```

I was found on image from 20241011.

**Comment by tfogal** [comment]
@vedaanta this is now the thirdmost important error in terms of killing mixology runs, preventing us from understanding perf for those models. Can this be prioritized in the near term?

**Comment by vedaanta** [comment]
Thanks for the bumps, @mpatel31415 and @tfogal.
tldr; It seems not be a cudnn issue?
Can we get more datapoints which would point to cudnn being the culprit?

----- 

There are two images I am running with:
1. pjnl-20241011
2. pjnl-20241011 without cudnn frontend
A Dockerfile with below contents should be sufficient to not have thunder use cudnn executor.
```
FROM INTERNAL:pjnl-20241011

RUN pip uninstall -y nvidia-cudnn-frontend
```

I tried out multiple models with both setups of single node and 2 node runs.

With 2 node runs:
- I can reproduce the segmentation fault with pjnl-20241011, with and without cudnn.
- It happens with executors other than thunder_cudnn too
    - cudnn is a default executor is thunder now, so specifying thunder_cudnn as done in this bug description should be redundant?
- When dumping cudnn logs without cudnn present in the container, I get nothing. This ensures that no other executor (maybe pytorch native) is using cudnn internally
- When using cudnn executor, I can see the that only sdpa operation is claimed by cudnn.
    - The logs seem fine. Nothing that would point to cudnn execution stopping suddenly.

With 1 node run:
- I was never able to reproduce the seg faults
- I have tried various combinations of model names, batch sizes
- My jobs either run into cuda OOM or finish successfully.
(This single node repro is not so important anyway as the 2-node repro is working fine.)

**Comment by tfogal** [comment]
> Thanks for the bumps, @mpatel31415 and @tfogal. tldr; It seems not be a cudnn issue? Can we get more datapoints which would point to cudnn being the culprit?

Thanks; looks like I misread back in July. This is an fp8 thing not a cudnn thing. Reassigning.

**Comment by mpatel31415** [comment]
For the most recent set of issues I used this script to reproduce the error:
```
#!/bin/bash
#SBATCH -A YOUR_ACCOUNT
#SBATCH -p batch
#SBATCH -J YOUR_JOB_NAME
#SBATCH -N 2
#SBATCH --ntasks-per-node 8
#SBATCH --time 0:29:00
#SBATCH --mail-type=FAIL
#SBATCH --exclusive


IMAGE="INTERNAL_IMAGE:pjnl-20241011"

TRAINING_SCRIPT=$(cat << 'EOF'
set -e
NVFUSER_DISABLE=multidevice python /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py \
    --max_iters 20 \
    --warmup_iters 5 \
    --model_name tiny-llama-1.1b \
    --distributed_mode ddp \
    --shard_mode None \
    --compile dynamo_thunder \
    --checkpoint_activations False \
    --low_precision_mode fp8-delayed-te \
    --micro_batch_size 20


EOF
)

srun --export=ALL --container-image=${IMAGE} bash -c "${TRAINING_SCRIPT}"
```

**Comment by kshitij12345** [comment]
As per offline discussion with @mpatel31415 , the latest run was successful without any segfaults.
